src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const changesArray = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      const { keysToRemove } = action;
+
+      currentState = { ...currentState };
+
+      for (const key of keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    changesArray.push({ ...currentState });
```

Comment:
Pushing a clone of `currentState` into `changesArray` ensures that each state change is captured accurately without being affected by subsequent actions.